OSDOCS-12627/OSDOCS-12728 Sigstore Support (#86087)
openshift/openshift-docs: modules/nodes-sigstore-configure-image-policy.adoc

// Module included in the following assemblies:
//
// * nodes/nodes-sigstore-using.adoc

:_mod-docs-content-type: PROCEDURE
[id="nodes-sigstore-configure-image-policy_{context}"]
= Creating an image policy CR

> An `ImagePolicy` custom resource (CR) enables a cluster administrator or application developer to configure a Sigstore signature verification policy for a specific namespace. The MCO watches `ImagePolicy` instance in different namespaces and merges the instances for each namespace into a `<namespace>.json` file in the `99-<pool>-generated-registries machine-config`. A cluster image policy overrides an image policy. 

Reviewer: ```suggestion
An `ImagePolicy` custom resource (CR) enables a cluster administrator or application developer to configure a Sigstore signature verification policy for a specific namespace. The MCO watches `ImagePolicy` instance in different namespaces and merges the instances for each namespace into a `/etc/crio/policies/<NAMESPACE>.json` file in the `99-<pool>-generated-registries machine-config`. A cluster image policy overrides an image policy. 
```

+
[source,terminal]
----
$ oc edit featuregate cluster
----
+
.Example `FeatureGate` CR
[source,yaml]
----
apiVersion: config.openshift.io/v1
kind: FeatureGate
metadata:
  name: cluster
spec:
  featureSet: TechPreviewNoUpgrade <1>
----
<1> Enables the required `SigstoreImageVerification` feature.
+
[WARNING]
====
Enabling the `TechPreviewNoUpgrade` feature set on your cluster cannot be undone and prevents minor version updates. This feature set allows you to enable these Technology Preview features on test clusters, where you can fully test them. Do not enable this feature set on production clusters.
====
+
After you save the changes, new machine configs are created, the machine config pools are updated, and scheduling on each node is disabled while the change is being applied.

.Procedure

. Create an image policy object similar to the following examples. See "About image policy parameters" for specifiec details on these parameters.
+
--
.Example image policy object with an exact repository match policy
[source,yaml]
----
apiVersion: config.openshift.io/v1alpha1
kind: ImagePolicy <1>
metadata:
  name: p0
  namespace: mynamespace <2>
spec:
  scopes: <3>

>     - example.com

Reviewer: ```suggestion
    - example.io/crio/signed
```

  policy: <4>
    rootOfTrust: <5>
      policyType: PublicKey <6>
      publicKey: 
        keyData: a2V5RGF0YQ== <7>
        rekorKeyData: cmVrb3JLZXlEYXRh <8>
    signedIdentity:
      matchPolicy: ExactRepository <9>
      exactRepository:

>         repository: example.com/repository/image <10>

Reviewer: ```suggestion
        repository: quay.io/crio/signed <10>
```

spec:
  scopes: <3>
    - example.com
  policy: <4>
    rootOfTrust: <5>
      policyType: FulcioCAWithReko <6>
      publicKey: <7>
        fulcioCAData: a2V5RGF0YQ==
        fulcioSubject:
          oidcIssuer: "\https://expected.OIDC.issuer/"
          signedEmail: "\expected-signing-user\@example.com"
        rekorKeyData: cmVrb3JLZXlEYXRh <8>
    signedIdentity:
      matchPolicy: exactRepository <9>
      exactRepository:
        repository: example.com/repository/image <10>  
----
<1> Creates an `ImagePolicy` object.
<2> Specifies the namespace where the image policy is applied. 
<3> Defines a list of image scopes assigned to this policy.
<4> Specifies the parameters that define how the images are verified.
<5> Defines a root of trust for the policy.
<6> Specifies the policy types that define the root of trust. Here, a Fulcio certification with required Rekor verification.
<7> For a Fulcio certificate policy, the following parameters are required:
* `fulcioCAData`: Specifies a base64-encoded Fulcio certificate in the PEM format. The maximum length is 8192 characters.
* `fulcioSubject`: Specifies the OIDC issuer and the email of the Fulcio authentication configuration.

> <8> Specifies a base64-encoded Rekor public key in the PEM format. This parameter is required when when the `policyType` is `FulcioCAWithReko`. The maximum length is 8192 characters.

Reviewer: ```suggestion
<8> Specifies a base64-encoded Rekor public key in the PEM format. This parameter is required when when the `policyType` is `FulcioCAWithRekor`. The maximum length is 8192 characters.
```

<9> Optional: Specifies one of the following processes to verify the identity in the signature and the actual image identity:
* `matchRepoDigestOrExact` 
* `matchRepository`.
* `exactRepository`
* `remapIdentity`
<10> For the `exactRepository` match policy, specifies the repository that contains the image identity and signature.
--

> . Create the cluster image policy object:

Reviewer: ```suggestion
. Create the image policy object:
```

+
[source,terminal]
----

> $ oc create -f <file-name>.yaml

Reviewer: I am not sure if this refers to the PublicKey or the FulcioCAWithRekor example above. I reviewed the example output below, assuming it refers to the PublicKey ImagePolicy.

----
+
Scheduling is disabled on each node as it is updated.

.Verification

. After the nodes in your cluster are updated, you can verify that the image policy has been configured:

> .. Enter into the node in debug mode by running the following command::

Reviewer: ```suggestion
.. Enter into the node in debug mode by running the following command:
```

+
[source,terminal]
----
$ oc debug node/<node_name>
----

.. Set `/host` as the root directory within the debug shell by running the following command:
+
[source,terminal]
----
sh-5.1# chroot /host/
----

.. Examine the `<namespace>.json` file  by running the following command: 
+
.Example output
[source,json]
----
sh-5.1# cat /etc/crio/policies/mynamespace.json
----
+
.Example output
[source,json]
----
# ...
  "docker": {

>    "example.com": [

Reviewer: ```suggestion
   "example.io/crio/signed": [
```

    {
     "type": "sigstoreSigned",
     "keyData": "a2V5RGF0YQ==",
     "rekorPublicKeyData": "cmVrb3JLZXlEYXRh",
     "signedIdentity": {

>       "type": "matchRepoDigestOrExact"

Reviewer: ```suggestion
      "type": "exactRepository"
      "dockerRepository": "quay.io/crio/signed"
```

This example will look like:
```
  "docker": {
   "example.io/crio/signed": [
    {
     "type": "sigstoreSigned",
     "keyData": "a2V5RGF0YQ==",
     "rekorPublicKeyData": "cmVrb3JLZXlEYXRh",
     "signedIdentity": {
      "type": "exactRepository",
      "dockerRepository": "quay.io/crio/signed"
     }
```

     }
# ...
----

////
..  Check the crio log for sigstore signature verification for namespaced policy by running the following command:
+
[source,terminal]
----

> sh-5.1#  journalctl -u crio | grep -A 100 "Pulling image: docker.io"

Reviewer: This log is for verifying image under docker.io. I think we'd better use quay.io and the above example should match the image specified in this log. 
I will take a look at to draft steps for this fulcio example.

Reviewer: FYI https://docs.google.com/document/d/10oZ1qEXBYI4TLxYwHez7054Q54KMXwo6p1xENPWhfvY/edit?usp=sharing
This is an example `FulcioCAWithRekor` verifying the image `example.io/crio/signed, signed identity is `quay.io/crio/signed`.
We can use the scope `example.io/crio/signed`

Author: @QiWang19 Does the user **_need_** to create a pod that pulls from one of the scope registries, or can we assume that users are configuring sigstore for repos that the pods are already pulling from?

> sh-5.1#  journalctl -u crio | grep -A 100 "Pulling image: docker.io"

Reviewer: ```suggestion
sh-5.1#  journalctl -u crio | grep -A 100 "Pulling image: example.io/crio"
```

Author: @QiWang19 I was not able to get this command to work for the Fulcio image policy example. Perhaps @lyman9966 Can help verify?

Reviewer: Since this policy for example.io/crio has signedIdentity configuration, it will need the ImageTagMirrorSet to redirect the image pull for `example.io/crio/signed` to `quay.io/crio/signed`, did you create the ImageTagMirrorSet as the doc: https://docs.google.com/document/d/10oZ1qEXBYI4TLxYwHez7054Q54KMXwo6p1xENPWhfvY/edit?usp=sharing?

```
"signedIdentity": {
      "type": "exactRepository",
      "dockerRepository": "quay.io/crio/signed"
     }
```

Author: @QiWang19 Ah. The ITMS worked. Thank you. 

> Since this policy for example.io/crio has signedIdentity configuration, it will need the ImageTagMirrorSet to redirect the image pull

This is because the scoped repo in this example is different than the signedIdentity repo? Having an ITMS isn't required for a signedIdentity configuration?
If so, maybe we should make the repos in doc examples the same  to make it easier. Maybe a note somewhere that you can use mirror registries (requiring an ITMS).

Reviewer: > If so, maybe we should make the repos in doc examples the same to make it easier

yes, I think this would be good. So in the above example we need to specify `signedIdentity` `MatchRepository`, and use the scope quay.io/crio/signed.
I draft the example here: https://docs.google.com/document/d/1CM1t9NhIgmECwWyq4DxT9ouNulVM2XAl6Y_ONcNfzXU/edit?usp=sharing

----
+
.Example output
[source,terminal]
----

> Dec 06 08:58:37 ip-10-0-94-101 crio[2289]: time="2024-12-06 08:58:37.704339479Z" level=debug msg="IsRunningImageAllowed for image docker:docker.io/myrepo/rhel8:latest" file="signature/policy_eval.go:274"
> Dec 06 08:58:37 ip-10-0-94-101 crio[2289]: time="2024-12-06 08:58:37.704371669Z" level=debug msg=" Using transport \"docker\" specific policy section \"docker.io\"" file="signature/policy_eval.go:150"
> Dec 06 08:58:37 ip-10-0-94-101 crio[2289]: time="2024-12-06 08:58:37.704426760Z" level=debug msg="Reading /var/lib/containers/sigstore/lyman9966/rhel8@sha256=fb51de62cfc8dfdec6a1d7288f75ad18c57ddaf249da7da0552ca8c881b8ee95/signature-1" file="docker/docker_image_src.go:545"
> Dec 06 08:58:37 ip-10-0-94-101 crio[2289]: time="2024-12-06 08:58:37.704732347Z" level=debug msg="Looking for sigstore attachments in docker.io/myrepo/rhel8:sha256-fb51de62cfc8dfdec6a1d7288f75ad18c57ddaf249da7da0552ca8c881b8ee95.sig" file="docker/docker_client.go:1138"
> Dec 06 08:58:37 ip-10-0-94-101 crio[2289]: time="2024-12-06 08:58:37.704776952Z" level=debug msg="GET https://registry-1.docker.io/v2/lyman9966/rhel8/manifests/sha256-fb51de62cfc8dfdec6a1d7288f75ad18c57ddaf249da7da0552ca8c881b8ee95.sig" file="docker/docker_client.go:617"
> Dec 06 08:58:37 ip-10-0-94-101 crio[2289]: time="2024-12-06 08:58:37.882270334Z" level=debug msg="Content-Type from manifest GET is \"application/vnd.oci.image.manifest.v1+json\"" file="docker/docker_client.go:989"
> Dec 06 08:58:37 ip-10-0-94-101 crio[2289]: time="2024-12-06 08:58:37.882464127Z" level=debug msg="Found a sigstore attachment manifest with 1 layers" file="docker/docker_image_src.go:639"

Reviewer: ```suggestion
Dec 23 18:33:41 ip-10-0-69-81 crio[2288]: time="2024-12-23 18:33:41.857452176Z" level=debug msg="IsRunningImageAllowed for image docker:example.io/crio/signed:latest" file="signature/policy_eval.go:274"
Dec 23 18:33:41 ip-10-0-69-81 crio[2288]: time="2024-12-23 18:33:41.857484676Z" level=debug msg=" Using transport \"docker\" specific policy section \"example.io/crio/signed\"" file="signature/policy_eval.go:150"
Dec 23 18:33:41 ip-10-0-69-81 crio[2288]: time="2024-12-23 18:33:41.857541608Z" level=debug msg="Reading /var/lib/containers/sigstore/crio/signed@sha256=18b42e8ea347780f35d979a829affa178593a8e31d90644466396e1187a07f3a/signature-1" file="docker/docker_image_src.go:545"
Dec 23 18:33:41 ip-10-0-69-81 crio[2288]: time="2024-12-23 18:33:41.857594149Z" level=debug msg="Looking for sigstore attachments in quay.io/crio/signed:sha256-18b42e8ea347780f35d979a829affa178593a8e31d90644466396e1187a07f3a.sig" file="docker/docker_client.go:1138"
Dec 23 18:33:41 ip-10-0-69-81 crio[2288]: time="2024-12-23 18:33:41.857634280Z" level=debug msg="GET https://quay.io/v2/crio/signed/manifests/sha256-18b42e8ea347780f35d979a829affa178593a8e31d90644466396e1187a07f3a.sig" file="docker/docker_client.go:617"
Dec 23 18:33:41 ip-10-0-69-81 crio[2288]: time="2024-12-23 18:33:41.931904818Z" level=debug msg="Content-Type from manifest GET is \"application/vnd.oci.image.manifest.v1+json\"" file="docker/docker_client.go:989"
Dec 23 18:33:41 ip-10-0-69-81 crio[2288]: time="2024-12-23 18:33:41.932149493Z" level=debug msg="Found a sigstore attachment manifest with 1 layers" file="docker/docker_image_src.go:639"
Dec 23 18:33:41 ip-10-0-69-81 crio[2288]: time="2024-12-23 18:33:41.932184524Z" level=debug msg="Fetching sigstore attachment 1/1: sha256:8276724a208087e73ae5d9d6e8f872f67808c08b0acdfdc73019278807197c45" file="docker/docker_image_src.go:644"
```

----
+

> You can see references to sigstore such as: `Looking for sigstore attachments` and `Found a sigstore attachment manifest with 1 layers`.

Reviewer: https://issues.redhat.com/browse/OCPNODE-2254?focusedId=26282787&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-26282787  This comment clarifies that the log message: `Using transport \"docker\" specific policy section...` indicates the policy.json configuration for Sigstore is being used to verify the signature.

> You can see references to sigstore such as: `Looking for sigstore attachments` and `Found a sigstore attachment manifest with 1 layers`.

Reviewer: ```suggestion
You can see references to sigstore such as: `Using transport \"docker\" specific policy section \"example.io/crio/signed\`.
```

--

. Create the image policy object:
+
[source,terminal]
----
$ oc create -f <file-name>.yaml
----
+
The Machine Config Operator (MCO) updates the machine config pools (MCP) in your cluster.

.Verification

. After the nodes in your cluster are updated, you can verify that the image policy has been configured:

.. Enter into the node in debug mode by running the following command:
+
[source,terminal]
----
$ oc debug node/<node_name>
----

.. Set `/host` as the root directory within the debug shell by running the following command:
+
[source,terminal]
----
sh-5.1# chroot /host/
----

.. Examine the `<namespace>.json` file by running the following command:
+
[source,terminal]
----
sh-5.1# cat /etc/crio/policies/<namespace>.json
----
+
.Example output for the image policy object with a public key showing the new image policy
[source,json]
----
# ...
 "transports": {
  "docker": {
   "example.io/crio/signed": [
    {
     "type": "sigstoreSigned",
     "keyData": "a2V5RGF0YQ==",
     "rekorPublicKeyData": "cmVrb3JLZXlEYXRh",
     "signedIdentity": {
      "type": "exactRepository",
      "dockerRepository": "quay.io/crio/signed"

>      }

Author: This is what I am seeing in the mynamespace.json file. 

![image](https://github.com/user-attachments/assets/2e957dac-fe90-4cf7-ab86-ad23861f1d89)

>      }

Reviewer: Ok, this looks good to me. There indentation difference from the code generating the file.

----
+
.Example output for the image policy object with a Fulcio certificate showing the new image policy
[source,json]
----
# ...
 "transports": {
  "docker": {

>    "example.io/crio/signed": [

Author: @QiWang19 Same as above.

>    "example.io/crio/signed": [

Reviewer: you can mark this as resolved.

[NOTE]
====
If a scoped image or repository in an image policy is nested under one of the scoped images or repositories in a cluster image policy, only the policy from cluster image policy is applied. However, the image policy object is created with an error message. For example, if an image policy specifies `example.com/global/image`, and the cluster image policy specifies `example.com/global`, the namespace inherits the policy from the cluster image policy.
====

The following example shows general guidelines on how to configure an `ImagePolicy` object. For more details on the parameters, see "About cluster and image policy parameters".

.Prerequisites
// Taken from https://issues.redhat.com/browse/OCPSTRAT-918
* You have a Sigstore-supported public key infrastructure (PKI), or provide link:https://docs.sigstore.dev/cosign/[Cosign public and private key pair] for signing operations.
* You have a signing process in place to sign your images.
* You have access to a registry that supports Cosign signatures.

> * You enabled the required Technology Preview features for your cluster by editing the `FeatureGate` CR named `cluster`:

Reviewer 2: Do we want to explicitly call out the need to mirror Sigstore signatures for OCP release images into your mirror registry before enabling Tech-Preview here?  Because if you don't, [the `openshift` ClusterImagePolicy][1] will block your ability to move the CVO Pod to new nodes, or to update to new OCP releases.

[1]: https://github.com/openshift/cluster-update-keys/blob/d44fca585d081a72cb2c67734556a27bbfc9470e/manifests.rhel/0000_90_openshift-cluster-image-policy.yaml

Reviewer: Mirroring Sigstore signatures for OCP release images, is this intended for disconnected users? 
I think we should also check if `oc-mirror` can be mentioned for signature mirroring guidance.

Reviewer 2: Without access to the Sigstore signatures, the current `openshift` ClusterImagePolicy will reject attempts to pull the release image.  But with [OCPSTRAT-1417](https://issues.redhat.com/browse/OCPSTRAT-1417) and [OCPSTRAT-1869](https://issues.redhat.com/browse/OCPSTRAT-1869) still undelivered, they can't use `oc-mirror` for that yet.  They'd have to use `oc image mirror ...` or other lower-level tooling.

Reviewer: Under this enabled the required Technology Preview bullet, add the following documentation:

Before enabling TechPreview, mirror Sigstore signatures for OCP release images into your mirror registry if mirrors are configured for OCP release image repositories. Otherwise, the openShift ClusterImagePolicy will block your ability to move the CVO Pod to new nodes or update to new OCP releases.
You can use `oc image mirror` to mirror the signatures. For example:
```
oc image mirror source.com/image/repo:sha256-1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef.sig \
mirror.com/image/repo:sha256-1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef.sig
```

@wking @mburke5678 What do you think?

Author: > move the CVO Pod to new nodes

Is this a common thing to do? I am unfamiliar with the CVO. But, I see in the docs that during an update, "the current CVO pod stops, and a CVO pod that uses the new version starts". Is this what you are referring to?

Reviewer: @wking do you think the above suggestion make sense https://github.com/openshift/openshift-docs/pull/86087#discussion_r1913802653? maybe we should not mention the mirror signature guidance since the oc-mirror is still undelivered

Reviewer 2: > > move the CVO Pod to new nodes
>
> Is this a common thing to do?

ClusterVersion updates will do this, yes.  But so will anything that rolls the control plane, e.g. configuring an additional mirror registry or adjusting the Proxy configuration, or otherwise creating a new, drain-inducing MachineConfig targeting the control-plane MachineConfigPool.  The MCO will drain the current CVO pod along with everything else from the node being updated, the Deployment controller will replace it with a new CVO Pod, the scheduler will put that CVO pod on a different control-plane Node (because the old one will still be cordoned and draining), and CRI-O will try to pull the new CVO Pod's requested image.  Once the control-plane nodes have the tech-preview ClusterImagePolicy in place, CRI-O will reject the CVO/release-image pull unless it can find the Red Hat signature alongside the release image in its configured mirrors.

Reviewer 2: [This][1]  bit:

> You can use oc image mirror to mirror the signatures. For example...

seems like a reasonable short-term approach to me, but your wording doesn't call out the importance of the OCP release image signature vs. the `openshift` ClusterImagePolicy, and "sorry, your CVO is dead unless you can get it back over to a Node that had already pulled that release image" seems like a high-stakes thing to surprise disconnected/restricted-network folks with.  But maybe we can find words to go along with the `oc image mirror ...` advice to make the importance of that particular signature clear?

[1]: https://github.com/openshift/openshift-docs/pull/86087#discussion_r1913802653

Author: @QiWang19 @wking Do I have this right?

> If registry mirrors are configured for the {product-title} release image repositories (`quay.io/openshift-release-dev/ocp-release` and `quay.io/openshift-release-dev/ocp-v4.0-art-dev`), before enabling the Technology Preview feature set, you must mirror the sigstore signatures for the {product-title} release images into your mirror registry. Otherwise, the default `openshift` cluster image policy blocks the ability of the Cluster Version Operator to move the CVO Pod to new nodes, which blocks the node update that results from the feature set change.

<7> For a Fulcio certificate policy, the following parameters are required:
* `fulcioCAData`: Specifies a base64-encoded Fulcio certificate in the PEM format. The maximum length is 8192 characters.
* `fulcioSubject`: Specifies the OIDC issuer and the email of the Fulcio authentication configuration.
<8> Specifies a base64-encoded Rekor public key in the PEM format. This parameter is required when when the `policyType` is `FulcioCAWithRekor`. The maximum length is 8192 characters.
<9> Optional: Specifies one of the following processes to verify the identity in the signature and the actual image identity:
* `MatchRepoDigestOrExact`.
* `MatchRepository`.
* `ExactRepository`. The `exactRepository` parameter must be specified.
* `RemapIdentity`. The `prefix` and `signedPrefix` parameters must be specified.
<10> For the `exactRepository` match policy, specifies the repository that contains the image identity and signature.
--

. Create the image policy object:
+
[source,terminal]
----
$ oc create -f <file-name>.yaml
----
+
The Machine Config Operator (MCO) updates the machine config pools (MCP) in your cluster.

.Verification

> . After the nodes in your cluster are updated, you can verify that the image policy has been configured:

Reviewer 3: Since this is a single step (with multiple sub-steps), IMO, a bullet can be used here.

[NOTE]
====
If a scoped image or repository in an image policy is nested under one of the scoped images or repositories in a cluster image policy, only the policy from cluster image policy is applied. However, the image policy object is created with an error message. For example, if an image policy specifies `example.com/global/image`, and the cluster image policy specifies `example.com/global`, the namespace inherits the policy from the cluster image policy.
====

The following example shows general guidelines on how to configure an `ImagePolicy` object. For more details on the parameters, see "About cluster and image policy parameters".

.Prerequisites
// Taken from https://issues.redhat.com/browse/OCPSTRAT-918
* You have a sigstore-supported public key infrastructure (PKI) or provide link:https://docs.sigstore.dev/cosign/[Cosign public and private key pair] for signing operations.
* You have a signing process in place to sign your images.
* You have access to a registry that supports Cosign signatures, if you are using Cosign signatures.

> * If registry mirrors are configured for the {product-title} release image repositories, `quay.io/openshift-release-dev/ocp-release` and `quay.io/openshift-release-dev/ocp-v4.0-art-dev`, before enabling the Technology Preview feature set, you must mirror the sigstore signatures for the {product-title} release images into your mirror registry. Otherwise, the default `openshift` cluster image policy blocks the ability of the Cluster Version Operator to move the CVO Pod to new nodes, which blocks the node update that results from the feature set change.

Reviewer: ```suggestion
* If registry mirrors are configured for the {product-title} release image repositories, `quay.io/openshift-release-dev/ocp-release` and `quay.io/openshift-release-dev/ocp-v4.0-art-dev`, before enabling the Technology Preview feature set, you must mirror the sigstore signatures for the {product-title} release images into your mirror registry. Otherwise, the default `openshift` cluster image policy that enforces signature verification for release repository will block the ability of the Cluster Version Operator to move the CVO Pod to new nodes, which blocks the node update that results from the feature set change.
```

+
You can use the `oc image mirror` command to mirror the signatures. For example:
+
[source,terminal]
----

> $ oc image mirror source.com/image/repo:sha256-1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef.sig \

Reviewer: ```suggestion
$ oc image mirror quay.io/openshift-release-dev/ocp-release:sha256-1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef.sig \
```

[id="nodes-sigstore-configure-image-policy_{context}"]
= Creating an image policy CR

An `ImagePolicy` custom resource (CR) enables a cluster administrator or application developer to configure a Sigstore signature verification policy for a specific namespace. The MCO watches `ImagePolicy` instances in different namespaces and updates the `/etc/crio/policies/<namespace>.json` and `/etc/containers/registries.d/sigstore-registries.yaml` files on all the nodes in the cluster.

[NOTE]
====
If a scoped image or repository in an image policy is nested under one of the scoped images or repositories in a cluster image policy, only the policy from cluster image policy is applied. However, the image policy object is created with an error message. For example, if an image policy specifies `example.com/global/image`, and the cluster image policy specifies `example.com/global`, the namespace inherits the policy from the cluster image policy.
====

> The following example shows general guidelines on how to configure an `ImagePolicy` object. For more details on the parameters, see "About cluster and image policy parameters".

Reviewer 4: The same with above: About cluster image policy and image policy parameters

sh-5.1# cat /etc/containers/registries.d/sigstore-registries.yaml
----
+
.Example output showing that the scoped registry was added
[source,yaml]
----
docker:

>   example.com:

Reviewer 4: In this example, the repo should be: example.io/crio/signed

Reviewer: this is still outstanding
